Spark on Kubernetes - basic submission client

liyinan926 opened the pull request:
Second draft upstreaming PR that contains the basic submission client implementation and unit tests. Branch `spark-kubernetes-3-updated` is a clone of `spark-kubernetes-3` with latest changes from `upstream/master` merged in. `spark-kubernetes-4` includes all our changes in `spark-kubernetes-3`.

cc @foxish @mccheah @apache-spark-on-k8s/contributors

foxish commented:
This will be the follow-up to https://github.com/apache/spark/pull/19468

mccheah commented:
This seems like a large diff, but a quick scan shows everything included as necessary. We need the driver service bootstrap because of changes to master. I think we can reduce the fanciness of the credentials step but that doesn't reduce the complexity by a significant amount.

foxish commented:
One TODO: Add the unit test in https://github.com/apache-spark-on-k8s/spark/pull/542 to this PR

liyinan926 commented:
Changed from #542 merged in.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,218 @@
+private[spark] object ClientArguments {
+  def fromCommandLineArgs(args: Array[String]): ClientArguments = {
+    var mainAppResource: Option[MainAppResource] = None
+    val otherPyFiles = Seq.empty[String]
```
Don't think we're using this here.

liyinan926 replied:
Removed.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,218 @@
+          addDriverOwnerReference(createdDriverPod, otherKubernetesResources)
+          kubernetesClient.resourceList(otherKubernetesResources: _*).createOrReplace()
+        }
+      } catch {
```
Don't think we want to catch Throwable here - look into `NonFatal`.

liyinan926 replied:
Changed to `NonFatal(e)`.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/BaseDriverConfigurationStep.scala:
```diff
@@ -0,0 +1,161 @@
+      .addAllToEnv(driverCustomEnvs.asJava)
+      .addToEnv(driverExtraClasspathEnv.toSeq: _*)
+      .addNewEnv()
+        .withName(ENV_DRIVER_MEMORY)
```
These environment variable keys don't make much sense without the DockerFile which describes the contract the submission client must fulfill.

It might make sense to include the Dockerfile for the driver and the executor in this PR. We shouldn't add the poms that build them - that would make this diff unnecessarily large.

liyinan926 replied:
Should we still put them under `docker-minimum-bundle`?

mccheah replied:
Yes. Though, for awhile now I've been thinking that there's probably a better name for this submodule =)

liyinan926 replied:
Should we just simply call it `docker`?

mccheah replied:
Think that should be fine.

liyinan926 replied:
Done.

mccheah commented:
A few comments but otherwise this captures the spirit of what we want to have upstream.

liyinan926 commented on resource-managers/kubernetes/docker/src/main/spark-base/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+COPY bin /opt/spark/bin
+COPY sbin /opt/spark/sbin
+COPY conf /opt/spark/conf
+COPY dockerfiles/spark-base/entrypoint.sh /opt/
```
@mccheah @foxish `dockerfiles/spark-base` doesn't make sense if the distribution does not include spark-k8s stuffs. Does this matter?

mccheah replied:
Think we still want to include it to show the projected contents of the image.

mccheah commented on resource-managers/kubernetes/docker/src/main/spark-base/Dockerfile:
```diff
@@ -0,0 +1,43 @@
+#
```
Think we want `src/main/dockerfiles` as the top level directory.

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/config.scala:
```diff
@@ -103,11 +129,45 @@ package object config extends Logging {
+      masterWithoutK8sPrefix
+    } else {
+      val resolvedURL = s"https://$masterWithoutK8sPrefix"
+      logDebug(s"No scheme specified for kubernetes master URL, so defaulting to https. Resolved" +
```
do not `s"` plain text...

felixcheung replied:
perhaps logInfo? this sounds useful

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigurationStepsOrchestrator.scala:
```diff
@@ -0,0 +1,85 @@
+      submissionSparkConf,
+      KUBERNETES_DRIVER_LABEL_PREFIX,
+      "label")
+    require(!driverCustomLabels.contains(SPARK_APP_ID_LABEL), s"Label with key " +
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/DriverConfigurationStepsOrchestrator.scala:
```diff
@@ -0,0 +1,85 @@
+      "label")
+    require(!driverCustomLabels.contains(SPARK_APP_ID_LABEL), s"Label with key " +
+      s" $SPARK_APP_ID_LABEL is not allowed as it is reserved for Spark bookkeeping" +
+      s" operations.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/BaseDriverConfigurationStep.scala:
```diff
@@ -0,0 +1,161 @@
+        "annotation")
+    require(!driverCustomAnnotations.contains(SPARK_APP_NAME_ANNOTATION),
+      s"Annotation with key $SPARK_APP_NAME_ANNOTATION is not allowed as it is reserved for" +
+        s" Spark bookkeeping operations.")
```
do not `s"` plain text...
and align this with the previous line

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverKubernetesCredentialsStep.scala:
```diff
@@ -0,0 +1,243 @@
+        }
+    val caCertDataBase64 = safeFileConfToBase64(
+        s"$KUBERNETES_AUTH_DRIVER_CONF_PREFIX.$CA_CERT_FILE_CONF_SUFFIX",
+        s"Driver CA cert file provided at %s does not exist or is not a file.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverKubernetesCredentialsStep.scala:
```diff
@@ -0,0 +1,243 @@
+        s"Driver CA cert file provided at %s does not exist or is not a file.")
+    val clientKeyDataBase64 = safeFileConfToBase64(
+        s"$KUBERNETES_AUTH_DRIVER_CONF_PREFIX.$CLIENT_KEY_FILE_CONF_SUFFIX",
+        s"Driver client key file provided at %s does not exist or is not a file.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverKubernetesCredentialsStep.scala:
```diff
@@ -0,0 +1,243 @@
+        s"Driver client key file provided at %s does not exist or is not a file.")
+    val clientCertDataBase64 = safeFileConfToBase64(
+        s"$KUBERNETES_AUTH_DRIVER_CONF_PREFIX.$CLIENT_CERT_FILE_CONF_SUFFIX",
+        s"Driver client cert file provided at %s does not exist or is not a file.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStep.scala:
```diff
@@ -0,0 +1,102 @@
+  override def configureDriver(driverSpec: KubernetesDriverSpec): KubernetesDriverSpec = {
+    require(submissionSparkConf.getOption(DRIVER_BIND_ADDRESS_KEY).isEmpty,
+      s"$DRIVER_BIND_ADDRESS_KEY is not supported in Kubernetes mode, as the driver's bind" +
+        s" address is managed and set to the driver pod's IP address.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStep.scala:
```diff
@@ -0,0 +1,102 @@
+        s" address is managed and set to the driver pod's IP address.")
+    require(submissionSparkConf.getOption(DRIVER_HOST_KEY).isEmpty,
+      s"$DRIVER_HOST_KEY is not supported in Kubernetes mode, as the driver's hostname will be" +
+        s" managed via a Kubernetes service.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStep.scala:
```diff
@@ -0,0 +1,102 @@
+      val shorterServiceName = s"spark-$randomServiceId$DRIVER_SVC_POSTFIX"
+      logWarning(s"Driver's hostname would preferably be $preferredServiceName, but this is" +
+        s" too long (must be <= 63 characters). Falling back to use $shorterServiceName" +
+        s" as the driver service's name.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala:
```diff
@@ -0,0 +1,234 @@
+      "-Dspark.logConf=true",
+      s"-D${SecondTestConfigurationStep.sparkConfKey}=" +
+        s"${SecondTestConfigurationStep.sparkConfValue}",
+      s"-XX:+HeapDumpOnOutOfMemoryError",
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala:
```diff
@@ -0,0 +1,234 @@
+      s"-D${SecondTestConfigurationStep.sparkConfKey}=" +
+        s"${SecondTestConfigurationStep.sparkConfValue}",
+      s"-XX:+HeapDumpOnOutOfMemoryError",
+      s"-XX:+PrintGCDetails")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStepSuite.scala:
```diff
@@ -0,0 +1,180 @@
+      case e: Throwable =>
+        assert(e.getMessage ===
+          s"requirement failed: ${DriverServiceBootstrapStep.DRIVER_BIND_ADDRESS_KEY} is" +
+            s" not supported in Kubernetes mode, as the driver's bind address is managed" +
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStepSuite.scala:
```diff
@@ -0,0 +1,180 @@
+        assert(e.getMessage ===
+          s"requirement failed: ${DriverServiceBootstrapStep.DRIVER_BIND_ADDRESS_KEY} is" +
+            s" not supported in Kubernetes mode, as the driver's bind address is managed" +
+            s" and set to the driver pod's IP address.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStepSuite.scala:
```diff
@@ -0,0 +1,180 @@
+      case e: Throwable =>
+        assert(e.getMessage ===
+          s"requirement failed: ${DriverServiceBootstrapStep.DRIVER_HOST_KEY} is" +
+            s" not supported in Kubernetes mode, as the driver's hostname will be managed via" +
```
do not `s"` plain text...
and please align this

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/steps/DriverServiceBootstrapStepSuite.scala:
```diff
@@ -0,0 +1,180 @@
+        assert(e.getMessage ===
+          s"requirement failed: ${DriverServiceBootstrapStep.DRIVER_HOST_KEY} is" +
+            s" not supported in Kubernetes mode, as the driver's hostname will be managed via" +
+            s" a Kubernetes service.")
```
do not `s"` plain text...

liyinan926 replied:
Done.

felixcheung commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -590,6 +600,11 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
+        |
+        | Kubernetes only:
+        |  --kubernetes-namespace NS   The namespace in the Kubernetes cluster within which the
+        |                              spark application must be launched. The namespace must
```
seems like we say `the application` or `your application` here

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import io.fabric8.kubernetes.api.model.{ContainerBuilder, EnvVar, EnvVarBuilder, HasMetadata, OwnerReferenceBuilder, Pod, PodBuilder}
```
more than 6, use import ._

liyinan926 replied:
Done.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+    val resolvedDriverPod = new PodBuilder(currentDriverSpec.driverPod)
+      .editSpec()
+        .addToContainers(resolvedDriverContainer)
+        .endSpec()
```
align these?

liyinan926 replied:
The indention probably makes sense as `editSpec` returns an object different than `PodBuilder`.

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/BaseDriverConfigurationStep.scala:
```diff
@@ -0,0 +1,161 @@
+          .build())
+        .endEnv()
+      .withNewResources()
+        .addToRequests("cpu", driverCpuQuantity)
```
what happens when we require for more than what the agent has? eg. more cores than number of cores in the agent node?

liyinan926 replied:
This will cause the driver to not be scheduled by the k8s scheduler onto a node until a node with that many cores becomes available in the cluster.

felixcheung replied:
In general YARN doesn't handle this case nicely either, but I wonder in k8s we could do better if it is hooked up to an autoscaler or something?

felixcheung commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/steps/BaseDriverConfigurationStep.scala:
```diff
@@ -0,0 +1,161 @@
+        .addToAnnotations(allDriverAnnotations.asJava)
+      .endMetadata()
+      .withNewSpec()
+        .withRestartPolicy("Never")
```
this might be configurable in the future? some sort of driver HA in cluster mode?

liyinan926 replied:
Good point. I think we can make it configurable.

foxish replied:
We should have an issue to track this. HA driver would be useful in a streaming context in the future.

liyinan926 replied:
https://github.com/apache-spark-on-k8s/spark/issues/546

felixcheung replied:
+1000x

felixcheung commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/k8s/submit/ClientSuite.scala:
```diff
@@ -0,0 +1,234 @@
+import org.apache.spark.deploy.k8s.constants._
+import org.apache.spark.deploy.k8s.submit.steps.{DriverConfigurationStep, KubernetesDriverSpec}
+
+private[spark] class ClientSuite extends SparkFunSuite with BeforeAndAfter {
```
don't do private[spark] with the Suite classes - it won't run by jenkins
see https://github.com/apache-spark-on-k8s/spark/pull/498/commits/c052212888e01eac90a006bfb5d14c513e33d0a3

liyinan926 replied:
Done.

liyinan926 commented:
If there's no objection, I will squash the commits and push to upstream for review by EOD today. @apache-spark-on-k8s/contributors

foxish commented:
SGTM! We should see how we can make it less confusing for reviewers - because this PR encompasses changes in spark-kubernetes-3.

liyinan926 commented:
When pushing upstream, I'm gonna remove code for the first PR so this is less confusing.

kimoonkim commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -296,6 +298,14 @@ object SparkSubmit extends CommandLineUtils with Logging {
+      case (KUBERNETES, CLIENT) =>
+        printErrorAndExit("Client mode is currently not supported for Kubernetes.")
+      case (KUBERNETES, CLUSTER) if args.isPython =>
+        printErrorAndExit("Cluster deploy mode is currently not supported for python " +
```
I wonder if this message could mislead users to think python client mode is supported. Users will try python client mode and will get the error in line 302, which is not a good experience. Is it possible to pattern match `case (KUBERNEETES, _) if args.isPython` instead, before line 301? Then, the error message can say "python is not supported for Kubernetes".

liyinan926 replied:
Done.

kimoonkim commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -296,6 +298,14 @@ object SparkSubmit extends CommandLineUtils with Logging {
+      case (KUBERNETES, CLUSTER) if args.isPython =>
+        printErrorAndExit("Cluster deploy mode is currently not supported for python " +
+          "applications on Kubernetes clusters.")
+      case (KUBERNETES, CLUSTER) if args.isR =>
```
Ditto. Wonder if we can match `case (KUBERNETES, _) if args.isR` before line 301 and say "R is not supported in Kubernetes".

liyinan926 replied:
Done.

kimoonkim commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -702,6 +717,16 @@ object SparkSubmit extends CommandLineUtils with Logging {
+      childMainClass = "org.apache.spark.deploy.k8s.submit.Client"
+      childArgs ++= Array("--primary-java-resource", args.primaryResource)
+      childArgs ++= Array("--main-class", args.mainClass)
+      args.childArgs.foreach { arg =>
```
I see a null check on `args.childArgs` is done at line 695 and a few others:
```
 if (args.childArgs != null) {
     args.childArgs.foreach { arg => childArgs += ("--arg", arg) }
 }    
```
Maybe we should do the same?

liyinan926 replied:
Done.

kimoonkim commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmitArguments.scala:
```diff
@@ -466,6 +473,9 @@ private[deploy] class SparkSubmitArguments(args: Seq[String], env: Map[String, S
       case USAGE_ERROR =>
         printUsageAndExit(1)
 
+      case KUBERNETES_NAMESPACE =>
```
Nit. I would expect a significant parameter like this to come before line 464. Line 464 - 473 are mostly about help and usage errors, it seems.

liyinan926 replied:
Done.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+     mainClass: String,
+     driverArgs: Array[String])
+
+private[spark] object ClientArguments {
```
I personally prefer one empty line between class/object header and the body, but I don't know if this is standard.

liyinan926 replied:
Done.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+
+    val resolvedDriverJavaOpts = currentDriverSpec
+      .driverSparkConf
+      // We don't need this anymore since we just set the JVM options on the environment
```
Where is the code that this comment is referring to?

liyinan926 replied:
Good catch. This comment and the line below it should be removed. Cc @foxish @mccheah to confirm.

liyinan926 replied:
Actually it seems it was added in #365. I think we should keep the comment but rephrase it to make it clearer.

mccheah replied:
Comment still applies because the lines below take the driver JVM options and append them to the JVM options of the SparkConf. For example, we don't want to have SPARK_DRIVER_JAVA_OPTS have a value of `-Dspark.driver.extraJavaOptions=-XX:HeapDumpOnOutOfMemoryError` - we just want the `-XX:HeapDumpOn...` to be set directly on the driver process.

foxish replied:
The comment should be rephrased to clarify the intent of removing `extraJavaOptions`, along the lines of what Matt said.

liyinan926 replied:
Done.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+}
+
+private[spark] object Client {
+  def run(sparkConf: SparkConf, clientArguments: ClientArguments): Unit = {
```
Same comment about one empty line before the class body.

liyinan926 replied:
Done.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+    val waitForAppCompletion = sparkConf.get(WAIT_FOR_APP_COMPLETION)
+    val appName = sparkConf.getOption("spark.app.name").getOrElse("spark")
+    val master = getK8sMasterUrl(sparkConf.get("spark.master"))
+    val loggingInterval = Option(sparkConf.get(REPORT_INTERVAL)).filter( _ => waitForAppCompletion)
```
Nit. Don't need a space before `_`?

liyinan926 replied:
Done.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,215 @@
+      clientArguments.driverArgs,
+      sparkConf)
+
+    Utils.tryWithResource(SparkKubernetesClientFactory.createKubernetesClient(
```
This particular line does not read well because the word "KubernetesClient" appears here twice meaning two different things. The reader may fail to distinguish "Spark client" (`SparkKubernetesClientFactory`), vs "K8s API client" (`createKubernetesClient`).

liyinan926 replied:
Renamed `SparkKubernetesClientFactory` to `KubernetesClientFactory` and renamed the method to `create`.

kimoonkim replied:
I see. I myself misread. The two clients meant the same thing :-)

kimoonkim reviewed (commented):
Initial batch of comments, mostly about readability. I still need to look at the various submission/configuration steps.

kimoonkim commented:
The latest commit seems to address my comments so far. Thanks!

liyinan926 committed:
Spark on Kubernetes - basic submission client

liyinan926 commented:
Squashed the commit and removed scheduler backend code and relevant changes in Yarn-related code.

liyinan926 commented:
@kimoonkim any more comments on the submission steps?

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/ConfigurationUtils.scala:
```diff
@@ -0,0 +1,49 @@
+
+import org.apache.spark.SparkConf
+
+private[spark] object ConfigurationUtils {
```
Put an empty line before the body?

liyinan926 replied:
Done.

liyinan926 commented:
This is under review at https://github.com/apache/spark/pull/19717.

kimoonkim commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/Client.scala:
```diff
@@ -0,0 +1,218 @@
+          kubernetesClient,
+          waitForAppCompletion,
+          appName,
+          loggingPodStatusWatcher).run()
```
I almost missed `.run()` at the end. Maybe we can use a`val` here:
```
kubernetesClient =>
        val sparkClient = new Client(
           configurationStepsOrchestrator.getAllConfigurationSteps(),
           sparkConf,
           kubernetesClient,
           waitForAppCompletion,
           appName,
           loggingPodStatusWatcher)
        sparkClient.run()
```

liyinan926 replied:
Done.

kimoonkim reviewed (commented):
Submission steps look good to me. Left a few minor comments. PTAL.

liyinan926 committed:
Addressed first round of review comments

liyinan926 committed:
Made Client implement the SparkApplication trait

liyinan926 committed:
Addressed the second round of comments

foxish commented:
@ifilonenko @erikerlandson @mccheah Can you please help with integration tests here? We need signals on what's working and what's not working. We expect some integration tests to fail, but most should pass. @liyinan926 just informed me that we were missing some critical functionality with `local://` file support. We want to ensure issues like that don't creep into our upstream PR.

foxish commented:
cc @kimoonkim

liyinan926 commented:
To elaborate, we are missing `DependencyResolutionStep` in this PR. Without it, users won't even be able to use `local://` dependencies.

erikerlandson commented:
we may be having a problem here if the docker poms were elided - the integration testing can't build images to run its tests, IIUC

foxish commented:
Tracking in https://github.com/apache-spark-on-k8s/spark/issues/568. @erikerlandson, you're right. We need to separate out the integration testing now, so we can actually run it against any distribution - irrespective of whether it has the dockerfile target or not.

liyinan926 committed:
Added missing step for supporting local:// dependencies and addressed more comments

kiszk committed:
[SPARK-22570][SQL] Avoid to create a lot of global variables by using a local variable with allocation of an object in generated code

## What changes were proposed in this pull request?

This PR reduces # of global variables in generated code by replacing a global variable with a local variable with an allocation of an object every time. When a lot of global variables were generated, the generated code may meet 64K constant pool limit.
This PR reduces # of generated global variables in the following three operations:
* `Cast` with String to primitive byte/short/int/long
* `RegExpReplace`
* `CreateArray`

I intentionally leave [this part](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala#L595-L603). This is because this variable keeps a class that is dynamically generated. In other word, it is not possible to reuse one class.

## How was this patch tested?

Added test cases

Author: Kazuaki Ishizaki <ishizaki@jp.ibm.com>

Closes #19797 from kiszk/SPARK-22570.

liyinan926 committed:
Fixed Scala style check errors

Commit:
[SPARK-21417][SQL] Infer join conditions using propagated constraints

## What changes were proposed in this pull request?

This PR adds an optimization rule that infers join conditions using propagated constraints.

For instance, if there is a join, where the left relation has 'a = 1' and the right relation has 'b = 1', then the rule infers 'a = b' as a join predicate. Only semantically new predicates are appended to the existing join condition.

Refer to the corresponding ticket and tests for more details.

## How was this patch tested?

This patch comes with a new test suite to cover the implemented logic.

Author: aokolnychyi <anton.okolnychyi@sap.com>

Closes #18692 from aokolnychyi/spark-21417.

wangyum committed:
[SPARK-22489][SQL] Shouldn't change broadcast join buildSide if user clearly specified

## What changes were proposed in this pull request?

How to reproduce:
```scala
import org.apache.spark.sql.execution.joins.BroadcastHashJoinExec

spark.createDataFrame(Seq((1, "4"), (2, "2"))).toDF("key", "value").createTempView("table1")
spark.createDataFrame(Seq((1, "1"), (2, "2"))).toDF("key", "value").createTempView("table2")

val bl = sql("SELECT /*+ MAPJOIN(t1) */ * FROM table1 t1 JOIN table2 t2 ON t1.key = t2.key").queryExecution.executedPlan

println(bl.children.head.asInstanceOf[BroadcastHashJoinExec].buildSide)
```
The result is `BuildRight`, but should be `BuildLeft`. This PR fix this issue.
## How was this patch tested?

unit tests

Author: Yuming Wang <wgyumg@gmail.com>

Closes #19714 from wangyum/SPARK-22489.

adrian-ionescu committed:
[SPARK-22614] Dataset API: repartitionByRange(...)

## What changes were proposed in this pull request?

This PR introduces a way to explicitly range-partition a Dataset. So far, only round-robin and hash partitioning were possible via `df.repartition(...)`, but sometimes range partitioning might be desirable: e.g. when writing to disk, for better compression without the cost of global sort.

The current implementation piggybacks on the existing `RepartitionByExpression` `LogicalPlan` and simply adds the following logic: If its expressions are of type `SortOrder`, then it will do `RangePartitioning`; otherwise `HashPartitioning`. This was by far the least intrusive solution I could come up with.

## How was this patch tested?
Unit test for `RepartitionByExpression` changes, a test to ensure we're not changing the behavior of existing `.repartition()` and a few end-to-end tests in `DataFrameSuite`.

Author: Adrian Ionescu <adrian@databricks.com>

Closes #19828 from adrian-ionescu/repartitionByRange.

gaborgsomogyi committed:
[SPARK-22428][DOC] Add spark application garbage collector configurat…

## What changes were proposed in this pull request?

The spark properties for configuring the ContextCleaner are not documented in the official documentation at https://spark.apache.org/docs/latest/configuration.html#available-properties.

This PR adds the doc.

## How was this patch tested?

Manual.

```
cd docs
jekyll build
open _site/configuration.html
```

Author: gaborgsomogyi <gabor.g.somogyi@gmail.com>

Closes #19826 from gaborgsomogyi/SPARK-22428.

Victsm committed:
[SPARK-22373] Bump Janino dependency version to fix thread safety issue…

… with Janino when compiling generated code.

## What changes were proposed in this pull request?

Bump up Janino dependency version to fix thread safety issue during compiling generated code

## How was this patch tested?

Check https://issues.apache.org/jira/browse/SPARK-22373 for details.
Converted part of the code in CodeGenerator into a standalone application, so the issue can be consistently reproduced locally.
Verified that changing Janino dependency version resolved this issue.

Author: Min Shen <mshen@linkedin.com>

Closes #19839 from Victsm/SPARK-22373.

tgravescs committed:
[SPARK-22653] executorAddress registered in CoarseGrainedSchedulerBac…

https://issues.apache.org/jira/browse/SPARK-22653
executorRef.address can be null, pass the executorAddress which accounts for it being null a few lines above the fix.

Manually tested this patch. You can reproduce the issue by running a simple spark-shell in yarn client mode with dynamic allocation and request some executors up front. Let those executors idle timeout. Get a heap dump. Without this fix, you will see that addressToExecutorId still contains the ids, with the fix addressToExecutorId is properly cleaned up.

Author: Thomas Graves <tgraves@oath.com>

Closes #19850 from tgravescs/SPARK-22653.

sujith71955 committed:
[SPARK-22601][SQL] Data load is getting displayed successful on providing non existing nonlocal file path

## What changes were proposed in this pull request?
When user tries to load data with a non existing hdfs file path system is not validating it and the load command operation is getting successful.
This is misleading to the user. already there is a validation in the scenario of none existing local file path. This PR has added validation in the scenario of nonexisting hdfs file path
## How was this patch tested?
UT has been added for verifying the issue, also snapshots has been added after the verification in a spark yarn cluster

Author: sujith71955 <sujithchacko.2010@gmail.com>

Closes #19823 from sujith71955/master_LoadComand_Issue.

mpetruska committed:
[SPARK-22393][SPARK-SHELL] spark-shell can't find imported types in class constructors, extends clause

## What changes were proposed in this pull request?

[SPARK-22393](https://issues.apache.org/jira/browse/SPARK-22393)

## How was this patch tested?

With a new test case in `RepSuite`

----

This code is a retrofit of the Scala [SI-9881](https://github.com/scala/bug/issues/9881) bug fix, which never made it into the Scala 2.11 branches. Pushing these changes directly to the Scala repo is not practical (see: https://github.com/scala/scala/pull/6195).

Author: Mark Petruska <petruska.mark@gmail.com>

Closes #19846 from mpetruska/SPARK-22393.

zsxwing committed:
[SPARK-22638][SS] Use a separate queue for StreamingQueryListenerBus

## What changes were proposed in this pull request?

Use a separate Spark event queue for StreamingQueryListenerBus so that if there are many non-streaming events, streaming query listeners don't need to wait for other Spark listeners and can catch up.

## How was this patch tested?

Jenkins

Author: Shixiong Zhu <zsxwing@gmail.com>

Closes #19838 from zsxwing/SPARK-22638.

liyinan926 commented:
It seems that the `LocalDirectoryMountConfigurationStep` is also essential. @mccheah @foxish @kimoonkim @erikerlandson @ifilonenko to confirm. If yes, I will add it to this PR.

HyukjinKwon committed:
[SPARK-22651][PYTHON][ML] Prevent initiating multiple Hive clients for ImageSchema.readImages

## What changes were proposed in this pull request?

Calling `ImageSchema.readImages` multiple times as below in PySpark shell:

```python
from pyspark.ml.image import ImageSchema
data_path = 'data/mllib/images/kittens'
_ = ImageSchema.readImages(data_path, recursive=True, dropImageFailures=True).collect()
_ = ImageSchema.readImages(data_path, recursive=True, dropImageFailures=True).collect()
```

throws an error as below:

```
...
org.datanucleus.exceptions.NucleusDataStoreException: Unable to open a test connection to the given database. JDBC url = jdbc:derby:;databaseName=metastore_db;create=true, username = APP. Terminating connection pool (set lazyInit to true if you expect to start your database after your app). Original Exception: ------
java.sql.SQLException: Failed to start database 'metastore_db' with class loader org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1742f639f, see the next exception for details.
...
	at org.apache.derby.jdbc.AutoloadedDriver.connect(Unknown Source)
...
	at org.apache.hadoop.hive.metastore.HiveMetaStore.newRetryingHMSHandler(HiveMetaStore.java:5762)
...
	at org.apache.spark.sql.hive.client.HiveClientImpl.newState(HiveClientImpl.scala:180)
...
	at org.apache.spark.sql.hive.HiveExternalCatalog$$anonfun$databaseExists$1.apply$mcZ$sp(HiveExternalCatalog.scala:195)
	at org.apache.spark.sql.hive.HiveExternalCatalog$$anonfun$databaseExists$1.apply(HiveExternalCatalog.scala:195)
	at org.apache.spark.sql.hive.HiveExternalCatalog$$anonfun$databaseExists$1.apply(HiveExternalCatalog.scala:195)
	at org.apache.spark.sql.hive.HiveExternalCatalog.withClient(HiveExternalCatalog.scala:97)
	at org.apache.spark.sql.hive.HiveExternalCatalog.databaseExists(HiveExternalCatalog.scala:194)
	at org.apache.spark.sql.internal.SharedState.externalCatalog$lzycompute(SharedState.scala:100)
	at org.apache.spark.sql.internal.SharedState.externalCatalog(SharedState.scala:88)
	at org.apache.spark.sql.hive.HiveSessionStateBuilder.externalCatalog(HiveSessionStateBuilder.scala:39)
	at org.apache.spark.sql.hive.HiveSessionStateBuilder.catalog$lzycompute(HiveSessionStateBuilder.scala:54)
	at org.apache.spark.sql.hive.HiveSessionStateBuilder.catalog(HiveSessionStateBuilder.scala:52)
	at org.apache.spark.sql.hive.HiveSessionStateBuilder$$anon$1.<init>(HiveSessionStateBuilder.scala:69)
	at org.apache.spark.sql.hive.HiveSessionStateBuilder.analyzer(HiveSessionStateBuilder.scala:69)
	at org.apache.spark.sql.internal.BaseSessionStateBuilder$$anonfun$build$2.apply(BaseSessionStateBuilder.scala:293)
	at org.apache.spark.sql.internal.BaseSessionStateBuilder$$anonfun$build$2.apply(BaseSessionStateBuilder.scala:293)
	at org.apache.spark.sql.internal.SessionState.analyzer$lzycompute(SessionState.scala:79)
	at org.apache.spark.sql.internal.SessionState.analyzer(SessionState.scala:79)
	at org.apache.spark.sql.execution.QueryExecution.analyzed$lzycompute(QueryExecution.scala:70)
	at org.apache.spark.sql.execution.QueryExecution.analyzed(QueryExecution.scala:68)
	at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:51)
	at org.apache.spark.sql.Dataset$.ofRows(Dataset.scala:70)
	at org.apache.spark.sql.SparkSession.internalCreateDataFrame(SparkSession.scala:574)
	at org.apache.spark.sql.SparkSession.createDataFrame(SparkSession.scala:593)
	at org.apache.spark.sql.SparkSession.createDataFrame(SparkSession.scala:348)
	at org.apache.spark.sql.SparkSession.createDataFrame(SparkSession.scala:348)
	at org.apache.spark.ml.image.ImageSchema$$anonfun$readImages$2$$anonfun$apply$1.apply(ImageSchema.scala:253)
...
Caused by: ERROR XJ040: Failed to start database 'metastore_db' with class loader org.apache.spark.sql.hive.client.IsolatedClientLoader$$anon$1742f639f, see the next exception for details.
	at org.apache.derby.iapi.error.StandardException.newException(Unknown Source)
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.wrapArgsForTransportAcrossDRDA(Unknown Source)
	... 121 more
Caused by: ERROR XSDB6: Another instance of Derby may have already booted the database /.../spark/metastore_db.
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/ml/image.py", line 190, in readImages
    dropImageFailures, float(sampleRatio), seed)
  File "/.../spark/python/lib/py4j-0.10.6-src.zip/py4j/java_gateway.py", line 1160, in __call__
  File "/.../spark/python/pyspark/sql/utils.py", line 69, in deco
    raise AnalysisException(s.split(': ', 1)[1], stackTrace)
pyspark.sql.utils.AnalysisException: u'java.lang.RuntimeException: java.lang.RuntimeException: Unable to instantiate org.apache.hadoop.hive.ql.metadata.SessionHiveMetaStoreClient;'
```

Seems we better stick to `SparkSession.builder.getOrCreate()` like:

https://github.com/apache/spark/blob/51620e288b5e0a7fffc3899c9deadabace28e6d7/python/pyspark/sql/streaming.py#L329

https://github.com/apache/spark/blob/dc5d34d8dcd6526d1dfdac8606661561c7576a62/python/pyspark/sql/column.py#L541

https://github.com/apache/spark/blob/33d43bf1b6f55594187066f0e38ba3985fa2542b/python/pyspark/sql/readwriter.py#L105

## How was this patch tested?

This was tested as below in PySpark shell:

```python
from pyspark.ml.image import ImageSchema
data_path = 'data/mllib/images/kittens'
_ = ImageSchema.readImages(data_path, recursive=True, dropImageFailures=True).collect()
_ = ImageSchema.readImages(data_path, recursive=True, dropImageFailures=True).collect()
```

Author: hyukjinkwon <gurwls223@gmail.com>

Closes #19845 from HyukjinKwon/SPARK-22651.

srowen committed:
[SPARK-22634][BUILD] Update Bouncy Castle to 1.58

## What changes were proposed in this pull request?

Update Bouncy Castle to 1.58, and jets3t to 0.9.4 to (sort of) match.

## How was this patch tested?

Existing tests

Author: Sean Owen <sowen@cloudera.com>

Closes #19859 from srowen/SPARK-22634.

dongjoon-hyun committed:
[SPARK-20682][SPARK-15474][SPARK-21791] Add new ORCFileFormat based on ORC 1.4.1

## What changes were proposed in this pull request?

Since [SPARK-2883](https://issues.apache.org/jira/browse/SPARK-2883), Apache Spark supports Apache ORC inside `sql/hive` module with Hive dependency. This PR aims to add a new ORC data source inside `sql/core` and to replace the old ORC data source eventually. This PR resolves the following three issues.

- [SPARK-20682](https://issues.apache.org/jira/browse/SPARK-20682): Add new ORCFileFormat based on Apache ORC 1.4.1
- [SPARK-15474](https://issues.apache.org/jira/browse/SPARK-15474): ORC data source fails to write and read back empty dataframe
- [SPARK-21791](https://issues.apache.org/jira/browse/SPARK-21791): ORC should support column names with dot

## How was this patch tested?

Pass the Jenkins with the existing all tests and new tests for SPARK-15474 and SPARK-21791.

Author: Dongjoon Hyun <dongjoon@apache.org>
Author: Wenchen Fan <wenchen@databricks.com>

Closes #19651 from dongjoon-hyun/SPARK-20682.

mgaido91 committed:
[SPARK-22669][SQL] Avoid unnecessary function calls in code generation

## What changes were proposed in this pull request?

In many parts of the codebase for code generation, we are splitting the code to avoid exceptions due to the 64KB method size limit. This is generating a lot of methods which are called every time, even though sometime this is not needed. As pointed out here: https://github.com/apache/spark/pull/19752#discussion_r153081547, this is a not negligible overhead which can be avoided.

The PR applies the same approach used in #19752 also to the other places where this was feasible.

## How was this patch tested?

existing UTs.

Author: Marco Gaido <mgaido@hortonworks.com>

Closes #19860 from mgaido91/SPARK-22669.

wangyum committed:
[SPARK-22626][SQL] deals with wrong Hive's statistics (zero rowCount)

This pr to ensure that the Hive's statistics `totalSize` (or `rawDataSize`) > 0, `rowCount` also must be > 0. Otherwise may cause OOM when CBO is enabled.

unit tests

Author: Yuming Wang <wgyumg@gmail.com>

Closes #19831 from wangyum/SPARK-22626.

wangyum committed:
[SPARK-22489][DOC][FOLLOWUP] Update broadcast behavior changes in migration section

## What changes were proposed in this pull request?

Update broadcast behavior changes in migration section.

## How was this patch tested?

N/A

Author: Yuming Wang <wgyumg@gmail.com>

Closes #19858 from wangyum/SPARK-22489-migration.

mgaido91 committed:
[SPARK-22473][FOLLOWUP][TEST] Remove deprecated Date functions

## What changes were proposed in this pull request?

#19696 replaced the deprecated usages for `Date` and `Waiter`, but a few methods were missed. The PR fixes the forgotten deprecated usages.

## How was this patch tested?

existing UTs

Author: Marco Gaido <mgaido@hortonworks.com>

Closes #19875 from mgaido91/SPARK-22473_FOLLOWUP.

Commit:
[SPARK-22162] Executors and the driver should use consistent JobIDs in the RDD commit protocol

I have modified SparkHadoopWriter so that executors and the driver always use consistent JobIds during the hadoop commit. Before SPARK-18191, spark always used the rddId, it just incorrectly named the variable stageId. After SPARK-18191, it used the rddId as the jobId on the driver's side, and the stageId as the jobId on the executors' side. With this change executors and the driver will consistently uses rddId as the jobId. Also with this change, during the hadoop commit protocol spark uses  actual stageId to check whether a stage can be committed unlike before that  it was using executors' jobId to do this check.
In addition to the existing unit tests, a test has been added to check whether executors and the driver are using the same JobId. The test failed before this change and passed after applying this fix.

Author: Reza Safi <rezasafi@cloudera.com>

Closes #19848 from rezasafi/stagerddsimple.

liyinan926 committed:
Addressed another round of comments

Commit:
[SPARK-22372][CORE, YARN] Make cluster submission use SparkApplication.

The main goal of this change is to allow multiple cluster-mode
submissions from the same JVM, without having them end up with
mixed configuration. That is done by extending the SparkApplication
trait, and doing so was reasonably trivial for standalone and
mesos modes.

For YARN mode, there was a complication. YARN used a "SPARK_YARN_MODE"
system property to control behavior indirectly in a whole bunch of
places, mainly in the SparkHadoopUtil / YarnSparkHadoopUtil classes.
Most of the changes here are removing that.

Since we removed support for Hadoop 1.x, some methods that lived in
YarnSparkHadoopUtil can now live in SparkHadoopUtil. The remaining
methods don't need to be part of the class, and can be called directly
from the YarnSparkHadoopUtil object, so now there's a single
implementation of SparkHadoopUtil.

There were two places in the code that relied on  SPARK_YARN_MODE to
make decisions about YARN-specific functionality, and now explicitly check
the master from the configuration for that instead:

* fetching the external shuffle service port, which can come from the YARN
  configuration.

* propagation of the authentication secret using Hadoop credentials. This also
  was cleaned up a little to not need so many methods in `SparkHadoopUtil`.

With those out of the way, actually changing the YARN client
to extend SparkApplication was easy.

Tested with existing unit tests, and also by running YARN apps
with auth and kerberos both on and off in a real cluster.

Author: Marcelo Vanzin <vanzin@cloudera.com>

Closes #19631 from vanzin/SPARK-22372.

liyinan926 committed:
Rebased on master and added a constant val for the Client class

liyinan926 committed:
Addressed another major round of comments

liyinan926 committed:
Addressed one more round of comments

liyinan926 committed:
Removed mentioning of kubernetes-namespace

liyinan926 committed:
Fixed a couple of bugs found during manual tests

liyinan926 committed:
Guard against client mode in SparkContext

liyinan926 committed:
Added libc6-compat into the base docker image

liyinan926 committed:
Addressed latest comments

liyinan926 committed:
Addressed docs comments

liyinan926 committed:
Fixed a comment

liyinan926 committed:
Addressed latest comments

liyinan926 commented:
The PR has been merged upstream. Closing this.
